Show Tree-sitter grammars in grammar selector

Aerijo opened the pull request:
### Requirements for Adding, Changing, or Removing a Feature

* Fill out the template below. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
* The pull request must contribute a change that has been endorsed by the maintainer team. See details in the template below.
* The pull request must update the test suite to exercise the updated functionality. For guidance, please see https://flight-manual.atom.io/hacking-atom/sections/writing-specs/.
* After you create the pull request, all status checks must be pass before a maintainer reviews your contribution. For more details, please see https://github.com/atom/atom/tree/master/CONTRIBUTING.md#pull-requests.

### Issue or RFC Endorsed by Atom's Maintainers

<!--

Link to the issue or RFC that your change relates to. This must be one of the following:

* An open issue with the `help-wanted` label
* An open issue with the `triaged` label
* An RFC with "accepted" status

To contribute an enhancement that isn't covered by one of the items above, please follow our guide for suggesting an enhancement: https://github.com/atom/atom/blob/master/CONTRIBUTING.md#suggesting-enhancements

To contribute other changes, you must use a different template. You can see all templates at https://github.com/atom/atom/tree/master/.github/PULL_REQUEST_TEMPLATE.

-->

Fixes #17029 (I endorse it...)
Closes https://github.com/atom/language-html/issues/226

### Description of the Change

<!--

We must be able to understand the design of your change from this description. If we can't get a good idea of what the code will be doing from the description here, the pull request may be closed at the maintainers' discretion. Keep in mind that the maintainer reviewing this PR may not be familiar with or have worked with the code here recently, so please walk us through the concepts.

-->

Adds an optional argument to `atom.grammars.getGrammars` to allow Tree-sitter grammars as well. It defaults to false to avoid any backwards compatibility issues, though I'd be fine with making it true or removing it entirely, given that Tree-sitter is now on by default for widely used languages like JS, C, etc.

Tests have not been added yet. Still need to style the grammar selector and fix my Dev mode ideally.

### Alternate Designs

Aside from the above, the grammar selector element constructor also needs additional styling to mark the type of grammar. Leaving TextMate grammars as they are is fine and least confusing, but Tree-sitter grammars often share a name with these, so they look like duplicate entries. I was thinking of adding `Ts` before the scope, but I would ask @simurai or someone to show me how to do it (my attempt  failed miserably...).

### Possible Drawbacks

Current design - making the argument a direct parameter could be awkward if we decide to add more options in future. Making it an object would future proof it.

If we enable it by default, it may break some community packages that ... handle grammars? That aren't already somewhat broken if they are ignoring Tree-sitter anyway? I'm not convinced this is an issue, so would like some thoughts on it.

### Verification Process

Manually overrided it in my init script. Dev mode currently does not open for me, and I have not worked out why.

### Release Notes

<!--

Please describe the changes in a single line that explains this improvement in
terms that a user can understand.  This text will be used in Atom's release notes.

If this change is not user-facing or notable enough to be included in release notes
you may use the strings "Not applicable" or "N/A" here.

Examples:

- The GitHub package now allows you to add co-authors to commits.
- Fixed an issue where multiple cursors did not work in a file with a single line.
- Increased the performance of searching and replacing across a whole project.

-->

Show Tree-sitter grammars in grammar selector

Aerijo committed:
Return tree sitter grammars too

Aerijo committed:
Make including tree sitter grammars optional

Aerijo committed:
Merge branch 'master' of github.com:atom/atom

simurai commented:
> Leaving TextMate grammars as they are is fine and least confusing, but Tree-sitter grammars often share a name with these, so they look like duplicate entries. I was thinking of adding `Ts` before the scope

Or write `Ts` out to `Tree-sitter`?

![screen shot 2019-02-12 at 4 36 53 pm](https://user-images.githubusercontent.com/378023/52619270-947aae80-2ee4-11e9-809a-8ac3c2b2d20f.png)

Although I wonder... For users that don't follow Atom development or read the Atom blog, do they know what "Tree-sitter" means? We could also be more descriptive and say that there is a normal and an "advanced" version:

![screen shot 2019-02-12 at 4 37 23 pm](https://user-images.githubusercontent.com/378023/52619497-22569980-2ee5-11e9-8458-182cbd5854a9.png)

Aerijo commented:
@simurai Writing our the full Tree sitter is fine, but I don’t like calling it advanced; that would also be just as confusing IMO. They are already referred to as Tree-sitter in the core setting, and “advanced” is a generic word that could refer to anything, including potential future supported grammar engines.

E.g., if the Sublime grammar engine gets ported successfully (even a community package), there could end up be three types of grammar. 

It would be best if there was a reliable way to tell users about Tree-sitter vs TextMate grammars, and why they’d pick one over the other. Unfortunately, they’re unlikely to go to the grammar selector package README for this. Maybe a small description somewhere on the prompt could link to an explanation?

Aerijo commented:
Did not mean to close

50Wliu commented:
Is it possible for the Grammar Selector to hide grammars that also have a tree-sitter variant?

Aerijo commented:
@50Wliu Yes, there should be a setting for that (normal users don’t need TextMate, but there are still cases where I would want to specify the TextMate one without disabling things)

simurai commented:
Ohh.. It just occurred to me that `JavaScript (Tree-siter)` might also be confusing because it looks like it's some sort of "flavor" of JavaScript and not the parsing system used. How about having green badges that just say `Tree-sitter` in front of the scopes. I'll push that.

![screen shot 2019-02-13 at 2 05 13 pm](https://user-images.githubusercontent.com/378023/52688230-e2e98500-2f98-11e9-8e6d-7119ad16f12c.png)

> Maybe a small description somewhere on the prompt could link to an explanation?

That would be helpful. Not sure how to describe it in the very small space left. Without, users that are curious would have to google it and hopefully find a blog post like https://github.blog/2018-10-31-atoms-new-parsing-system/.

simurai committed:
Add “Tree-sitter” badge

simurai commented on packages/grammar-selector/lib/grammar-list-view.js:
```diff
@@ -19,9 +19,19 @@ class GrammarListView {
         div.classList.add('pull-right')
+
+        if (grammar.constructor.name === "TreeSitterGrammar") {
+          console.log("TS", grammar)
```
TODO: Remove before merging.

Arcanemagus commented:
> That would be helpful. Not sure how to describe it in the very small space left.

Can those `Tree-sitter` badges have alt-text added on hover?

simurai commented:
> Can those `Tree-sitter` badges have alt-text added on hover?

Yes, here with a `title` attribute:

![title](https://user-images.githubusercontent.com/378023/52693057-01577c80-2fa9-11e9-80f1-6546ddc645ef.gif)

Does the following description make sense?

```
A fast parsing system with improved syntax highlighting, code folding, syntax-aware selection.
```

Aerijo commented:
My suggestion:
```
(Recommended) A faster parser with improved syntax highlighting & code navigation support.
```
Shorter and, in case the user can't tell from the listed improvements, explicitly says it's recommended.

simurai committed:
Add description

simurai committed:
Remove console.log

Aerijo committed:
Merge branch 'master' of https://github.com/atom/atom

Aerijo commented:
OK, couple of outstanding issues I think can be fixed in this PR:
1. Order of TS and TM grammars is random when both have same name. E.g., C has TS first, but C++ has TM first.
2. Choosing TM grammar does not actually choose the TM grammar.

Aerijo committed:
Allow hiding duplicates + stable sort TS

Aerijo commented on packages/grammar-selector/lib/grammar-list-view.js:
```diff
@@ -20,7 +20,7 @@ class GrammarListView {
         div.classList.add('pull-right')
 
-        if (grammar.constructor.name === "TreeSitterGrammar") {
+        if (isTreeSitter(grammar)) {
```
This is used a few times, and as a function adds better separation of the idea from the implementation

Aerijo commented on packages/grammar-selector/lib/grammar-list-view.js:
```diff
@@ -82,32 +82,64 @@ class GrammarListView {
+      return
+    }
+
+    const editor = atom.workspace.getActiveTextEditor()
```
It's trivial, but the original felt wasteful compared to just keeping the editor the first time. It's also technically more correct perhaps? Can the editor change between two consecutive steps?

Aerijo commented on packages/grammar-selector/lib/grammar-list-view.js:
```diff
@@ -82,32 +82,64 @@ class GrammarListView {
       })
+
+      if (atom.config.get("grammar-selector.hideDuplicateTextMateGrammars")) {
+        const oldGrammars = grammars
```
This is a mess, but seems to work. Open for improvements.

Aerijo commented on packages/grammar-selector/lib/grammar-list-view.js:
```diff
@@ -82,32 +82,64 @@ class GrammarListView {
-          return a.scopeName.localeCompare(b.scopeName)
-        } else {
-          return 1
+        } else if (a.name === b.name) {
```
Equivalent logic (besides TS check), as we filtered out grammars with no name already.

Aerijo commented on packages/grammar-selector/lib/grammar-list-view.js:
```diff
@@ -82,32 +82,64 @@ class GrammarListView {
+  return grammar.constructor.name === "TreeSitterGrammar"
+}
+
+function compareGrammarType (a, b) {
```
Separate function to make extending to multiple grammar types easier and not bloat the main sort function

Aerijo commented on src/grammar-registry.js:
```diff
@@ -549,10 +549,12 @@ class GrammarRegistry {
-    let grammars = this.textmateRegistry.getGrammars()
-    if (includeTreesitter) grammars = grammars.concat(Object.values(this.treeSitterGrammarsById))
-    return grammars
+  getGrammars (params = {includeTreeSitter: false}) {
```
Turned into options object so we can add / change settings later without breaking changes

Maybe add TM-TS duplicate removal here? It would be slightly easier, and allow it to be used by all packages instead of making them do it themselves

Aerijo committed:
Make method to assign specific grammar

Aerijo committed:
Fix specs

Aerijo committed:
add specs

Aerijo commented:
Added specs for most behaviour (aren't checking the order of Tree-sitter / TextMate when both show).

The specs are quite rigid, expecting specific ordering and nesting of the DOM nodes, but they should not be flakey.

I also refactored getting the list node into it's own function for some quality-of-life improvement.

Aerijo committed:
merge upstream

Aerijo commented:
@simurai The button text is smaller than the label text
![image](https://user-images.githubusercontent.com/22167388/53713932-ac6c9f00-3ea0-11e9-92a5-86e7a8a1a3eb.png)

I'm using a different monitor now, and the high pixel density means app components look really tiny, especially text. In the image, the list items are 15px, but the badge text is 12px. Would it be possible to make the badge text inherit from the list by default?

Aerijo committed:
fix lints

Aerijo committed:
Better tests and enable returning Tree-sitter by default

Aerijo committed:
docs and minor tweaks

Aerijo committed:
fix specs

Aerijo committed:
fix lint

Aerijo committed:
fix serialisation spec

Aerijo committed:
actually fix it

Aerijo committed:
this?

rafeca committed:
Merge commit '677bbb7f0b8754787ff9e7bfab4602ba82e13b0b' into pr18738

rafeca committed:
Merge commit '1d9a4cafcf6cc288d675512db8fd984e13aab869' into pr18738

rafeca committed:
Reformat all JS files using prettier

rafeca commented:
Hey! 👋

We have recently started using [`prettier`](http://prettier.io/) and this has caused major style changes on the Atom JS codebase (you can check [the related PR](https://github.com/atom/atom/pull/19391)).

This is good news: now the Atom code is more consistent and it's much easier to re-format the code to follow the codestyle (now you only need to run `script/lint --fix`).

This change caused conflicts on your PR that we have automatically solved, hope you don't mind 😄

With ❤️, the Atom team.

Aerijo committed:
Merge branch 'master' of github.com:atom/atom

Aerijo committed:
go with default include approach

Aerijo committed:
fix lint

Aerijo commented:
I've been considering returning Tree-sitter grammars by default, or explicit opt in. I've gone with by default, because they are supposed to be the preferred grammar now. There were a couple of hitches with other parts of GrammarRegistry that expose grammars, but I think everything works (the specs pass).

The failing specs seem completely unrelated; the mac build failed to load, throwing
```
[command]/bin/bash --noprofile --norc /Users/vsts/agent/2.153.1/work/_temp/d77d45f0-4dd1-4d72-a48e-8ec869406f78.sh
internal/modules/cjs/loader.js:596
    throw err;
    ^

Error: Cannot find module 'colors'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:594:15)
    at Function.Module._load (internal/modules/cjs/loader.js:520:25)
    at Module.require (internal/modules/cjs/loader.js:650:17)
    at require (internal/modules/cjs/helpers.js:20:18)
    at Object.<anonymous> (/Users/vsts/agent/2.153.1/work/1/s/script/test:5:1)
    at Module._compile (internal/modules/cjs/loader.js:702:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:713:10)
    at Module.load (internal/modules/cjs/loader.js:612:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:551:12)
    at Function.Module._load (internal/modules/cjs/loader.js:543:3)
```

I think this is ready to be reviewed and tried in beta.

nathansobo commented on packages/grammar-selector/lib/grammar-list-view.js:
```diff
@@ -72,32 +88,68 @@ module.exports = class GrammarListView {
         return grammar !== atom.grammars.nullGrammar && grammar.name;
       });
+
+      if (atom.config.get('grammar-selector.hideDuplicateTextMateGrammars')) {
```
In this case, do you really need the second call to `atom.grammars.getGrammars({ textMateOnly: true }`? It seems like you could simply iterate through the grammars array twice, once to populate the blacklist of Tree Sitter grammars and then a second time to filter. It's not a huge deal though.

Aerijo replied:
Yeah, cleaned it up

nathansobo commented on src/grammar-registry.js:
```diff
@@ -601,9 +617,19 @@ module.exports = class GrammarRegistry {
   // Returns a non-empty {Array} of {Grammar} instances.
-  getGrammars() {
-    return this.textmateRegistry.getGrammars();
+  getGrammars(params) {
```
I think that if we want to return Tree Sitter grammars by default, which I agree would be the cleaner choice long term, then we need `TreeSitterGrammar` to conform to [the documented public API](https://atom.io/docs/api/v1.38.2/Grammar) of the `Grammar` objects we claim we are returning from this method. I think the `onDidUpdate` method could be a no-op, but we'd want a compatible implementation for `tokenizeLine` and `tokenizeLines`. If we return objects from this method that don't conform to that interface, we'll basically be invalidating our documented API. So I see two choices, both acceptable:

* Change this method to *not* return Tree Sitter grammars by default
* Backfill support for the public `Grammar` API on `TreeSitterGrammar` (this would actually be my preference)

Aerijo replied:
I've backfilled the public methods. `onDidUpdate` does nothing, and the `tokenizeLine`(s) functions return a generic result, applying only the root scope to each line.

nathansobo reviewed (changes requested):
I really love the UX you came up with that explains what a Tree Sitter grammar is to the user. Nice work. From a code perspective, this all seems legit. I'm concerned about the API implications however and have a comment with more details about it inline.

Aerijo committed:
clean up duplicate removal

Aerijo committed:
fix lint

Aerijo committed:
imitate public interface

Aerijo committed:
fix lint

Aerijo committed:
Merge branch 'master' of github.com:atom/atom

nathansobo commented:
The CI failure is really odd.

Commit:
Merge remote-tracking branch 'origin/master' into Aerijo

nathansobo commented:
Okay, I merged master into this branch after merging #19580, so hopefully we get better behavior out of the tests.

nathansobo commented:
Before we merge, I'm going to spend some time seeing if we can actually return tokens from the methods on grammar. The current shims won't *break* the current API, but I also think they don't really perform as we advertise in the documentation, which could be surprising. Let's see what I can come up with.

nathansobo commented:
Sorry. After digging into this I really don't think we can sanely shim `tokenizeLine`. It's intended to carry the state of the first-mate parser across calls, and my scan of all the package code out there shows that people are indeed using it. I think we should avoid returning objects that violate the documented interface of `Grammar` without an explicit option. And maybe we leave that option undocumented for now. I can take care of making the relevant changes.

Aerijo committed:
Default to no Tree-sitter grammars

Aerijo committed:
restore tree-sitter in grammar iterator

Aerijo committed:
adjust specs

Aerijo committed:
fix lint

Aerijo committed:
fix spec

Aerijo commented:
OK, it's now set back up to require opt in for Tree-sitter. The change took a few tries, because there are a couple of places grammars are exposed, and it wasn't consistent to begin with (see `forEachGrammar`, which exposed TS grammars for the auto assignment methods).

nathansobo commented:
Thanks so much! It will be really nice to be able to select these grammars via the UI. Really, the functionality was incomplete without this.

nathansobo commented:
A nice next step would be to start thinking about how we can expose the unique capabilities of TreeSitter grammars and language modes in our official API.

maxbrunsfeld commented:
Nice work @Aerijo and @simurai. Looks great.
